Add non-local networks section (#208)
iamdefinitelyahuman/brownie: docs/nonlocal-networks.rst

.. _nonlocal-networks:

====================

> Non-local Networks

Reviewer: ```suggestion
Using Non-Local Networks
```

====================

In addition to using `ganache-cli <https://github.com/trufflesuite/ganache-cli>`__ as a local development environment, Brownie can also connect to non-local networks (i.e. any testnet/mainnet node that supports JSON RPC).

Warning
========================

> Before you go any farther, consider that connecting to non-local networks can potentially expose your private keys if you aren't careful.

Reviewer: ```suggestion
Before you go any further, consider that connecting to non-local networks can potentially expose your private keys if you aren't careful.
```

* When you are interacting with mainnet, make sure you verify all of the details of any transactions you sign/send before you send them. Brownie can't protect you from sending ETH to the wrong address, sending too much, etc. 
* Always protect your private keys.  Don't leave them lying around unencrypted!

Register with Infura
========================
Before you can connect to a non-local network, you need access to a remote Ethereum node that supports JSON RPC over HTTP.  `Infura <https://infura.io>`__ is one such option.  Once you register and create a project, Infura will provide you URLs with the API key that can be leveraged to access the given network.

> 

Author: Updated section with guidance on setting environmental variables.

====================================

Using the CLI
-------------

Brownie will connect to whichever network is set as "default" in ``brownie-config.json``.  

To connect to any other network that has been defined in ``brownie-config.json``, you must specify the `--network` flag when launching Brownie as below:
::
    $ brownie --network ropsten

Each time Brownie is loaded, it will first attempt to connect to the ``host`` address to determine if the RPC client is already active.

Using the ``brownie.netowrk`` methods
-------------------------------------

You can interact with any network defined in ``brownie-config.json`` programatically using the ``brownie.network.main`` methods.

Connecting to a network:

.. code-block:: python

    >>> network.connect('ropsten')    
    >>> network.is_connected()
    True
    >>> network.show_active()
    'ropsten'

Disconnecting from a network:

.. code-block:: python

    >>> network.disconnect()
    >>> network.is_connected()
    False

Interacting with non-local networks
==============================

``brownie.module.rpc``
--------

The :ref:`rpc` module is unavailable when working with non-local networks.

Accounts
--------

> When loading an account for interacting with a non-local network, you must provide the private key when loading the account in order to be able to sign transactions or deploy contracts

Reviewer: Would be good to clarify here that ``Accounts`` already includes all unlocked addresses as returned by `eth_accounts`. If the user is connected to their own private node there is no need to use `add`, `save`, or `load` - these methods are mostly useful when working on a hosted node such as Infura.

Author: Clarification added.

.. code-block:: python

        >>> accounts.add('8fa2fdfb89003176a16b707fc860d0881da0d1d8248af210df12d37860996fb2')
        <Account object '0xc1826925377b4103cC92DeeCDF6F96A03142F37a'>
        >>> accounts[0].balance()
        17722750299000000000

Once an account is added to the account object, the ``accounts.save`` and ``accounts.load`` can be used to save the accounts to an encrypted keystore and then load for later use.

Transactions
------------

* Transaction status
When submitting a transaction on non-local networks, blocks are not immediately so transactions will likewise not be immediately confirmed. 

> A :ref:`api-network-tx` object is provided immediately and can be stored to unique variables though ``TransactionReceipt.status`` will be ``-1`` until the transaction is mined and either succeeds or reverts.  

Reviewer: A `TransactionReceipt` is not returned immediately by default, Brownie will display a "waiting for confirmation..." message until it confirms.  In the console, a user can press `CTRL-C` while waiting, and then they are immediately returned the pending `TransactionReceipt`.

Related: https://eth-brownie.readthedocs.io/en/latest/interaction.html#unconfirmed-transactions

Author: Clarifications added.

* Debugging 
The Brownie :ref:`debug` tools rely upon `debug_traceTransaction <https://github.com/ethereum/go-ethereum/wiki/Management-APIs#user-content-debug_tracetransaction>`__ RPC method which is not supported by `Infura <https://infura.io>`__. Attempts to call it will result in a ``RPCRequestError``.
This means that the below ``TransactionReceipt`` attributes and methods are unavailable:

* ``TransactionReceipt.return_value``
* ``TransactionReceipt.trace``
* ``TransactionReceipt.call_trace``
* ``TransactionReceipt.traceback``
* ``TransactionReceipt.source``

:ref:`api-network-contract`
---------

> On non-local networks, use the ``Contract`` class to interact with deployed contracts.  ``ContractContainer`` and ``ProjectContract`` are unavailable as these are only used with the local   

Reviewer: A project can be deployed to a non-local network, and then ``ProjectContract`` instances will be available for that network.

* Calling `ContractContainer.at` is useful for obtaining `ProjectContract` instances, to interact with deployed contracts that are part of the project
* Instantiating a `Contract` object is useful for interacting with contracts that are not of a project

Author: I've updated this section in line with this. Let me know if it's accurate.

* When you are interacting with mainnet, make sure you verify all of the details of any transactions you sign/send before you send them. Brownie can't protect you from sending ETH to the wrong address, sending too much, etc. 
* Always protect your private keys.  Don't leave them lying around unencrypted!

Register with Infura
========================
Before you can connect to a non-local network, you need access to an Ethereum node (whether your own local one or hosted) that supports JSON RPC (either HTTP, IPC, or web-sockets).  `Infura <https://infura.io>`__ is a good option for accessing a hosted node.  Once you register and create a project, Infura will provide you URLs with the API key that can be leveraged to access the given network.

Non-local Network Configuration
================================

Configuring non-local networks
------------------------------

The connection settings for non-local networks must be defined in ``brownie-config.json``.

First, for each network you want to configure, create a new section in the network.networks section as below:

.. code-block:: javascript

    "networks": {
        .
        .
        "ropsten": {

>             "host": "http://ropsten.infura.io/v3/[yourInfuraApiKey]"

Reviewer: ```suggestion
            "host": "http://ropsten.infura.io/v3/$WEB3_INFURA_PROJECT_ID"
```

        }
        "rinkeby":...
        .
        .
    }

> You can also set your Infura API key as an environment variable from the command line as below.  

Reviewer: ```suggestion
If using Infura, you can provide your project ID key as an environment variable or by modifying the ``hosts`` setting in the configuration file.

The environment variable is set to ``WEB3_INFURA_PROJECT_ID`` in the default configuration file.  Use the following command to set the environment variable:
```

::
    $ export WEB3_INFURA_PROJECT_ID=YourProjectID

Setting the default nerwork
---------------------------

> Brownie''s default configuration uses the environment variable WEB3_INFURA_PROJECT_ID.

Reviewer: ```suggestion
```

If you include the previous suggested edit, this line can be removed.

If you want to change the default network that brownie connects to, you need to update the network.default field as below:

.. code-block:: javascript

    "network": {
        "default": "ropsten",
        .
        .
    }

Launching and Connecting to Networks
====================================

Using the CLI
-------------

Brownie will connect to whichever network is set as "default" in ``brownie-config.json``.  

> To connect to any other network that has been defined in ``brownie-config.json``, you must specify the `--network` flag when launching Brownie as below:

Reviewer: ```suggestion
To connect to any other network that has been defined in ``brownie-config.json``, you must specify the ``--network`` flag when launching Brownie as below:
```

::
    $ brownie --network ropsten

Each time Brownie is loaded, it will first attempt to connect to the ``host`` address to determine if the RPC client is already active.

Using the ``brownie.netowrk`` methods
-------------------------------------

You can interact with any network defined in ``brownie-config.json`` programatically using the ``brownie.network.main`` methods.

Connecting to a network:

.. code-block:: python

    >>> network.connect('ropsten')    
    >>> network.is_connected()
    True
    >>> network.show_active()
    'ropsten'

Disconnecting from a network:

.. code-block:: python

    >>> network.disconnect()
    >>> network.is_connected()
    False

Interacting with non-local networks
==============================

``brownie.module.rpc``
--------

The :ref:`rpc` module is unavailable when working with non-local networks.

Accounts
--------

If you are connected to your own private node, Brownie will automatically load any unlocked accounts returned by your node.  In this case, there is no need to use ``accounts.load``.

> When interacting with a non-local network via a hosted node like Infura, you must provide the private key when loading your acccount in order to be able to sign transactions or deploy contracts

Reviewer: ```suggestion
When interacting with a non-local network via a hosted node such as Infura, you must provide the private key when loading your account in order to be able to sign transactions or deploy contracts.
```

.. code-block:: python

>         >>> accounts.add('8fa2fdfb89003176a16b707fc860d0881da0d1d8248af210df12d37860996fb2')

Reviewer: this looks over-indented

Author: Fixed in next commit.

        <Account object '0xc1826925377b4103cC92DeeCDF6F96A03142F37a'>
        >>> accounts[0].balance()
        17722750299000000000

Once an account is added to the account object, the ``accounts.save`` and ``accounts.load`` can be used to save the accounts to an encrypted keystore and then load for later use.

Transactions
------------

* Transaction status

> When submitting transactions on non-local networks, blocks are not immediately so transactions will likewise not be immediately confirmed. 

Reviewer: ```suggestion
When submitting transactions on non-local networks, blocks are not mined immediately so transactions will likewise not be immediately confirmed. 
```

Brownie does not provide a transaction receipt by default but will wait until the transaction has been confirmed before continuing execution.  
Press ``Ctrl-C`` and a :ref:`api-network-tx` object will be returned with the pending transaction hash and can be stored to unique variables. ``TransactionReceipt.status`` will be ``-1`` until the transaction is mined and either succeeds or reverts.  

* Debugging 
The Brownie :ref:`debug` tools rely upon `debug_traceTransaction <https://github.com/ethereum/go-ethereum/wiki/Management-APIs#user-content-debug_tracetransaction>`__ RPC method which is not supported by `Infura <https://infura.io>`__. Attempts to call it will result in a ``RPCRequestError``.
This means that the below ``TransactionReceipt`` attributes and methods are unavailable:

* ``TransactionReceipt.return_value``
* ``TransactionReceipt.trace``
* ``TransactionReceipt.call_trace``
* ``TransactionReceipt.traceback``
* ``TransactionReceipt.source``

:ref:`api-network-contract`
---------
Contracts
*********

On non-local networks, use the ``Contract`` class to interact with already deployed contracts.  

> You can instantiate the contract using ``contract.Contract`` method.  You will need to provide an ABI (typically as a JSON file) that is generated from the compiled contract code.  

Reviewer: `abi` is given as a `dict` - while true that it's typically generated from a JSON file, stating that it expectas a JSON could result in someone trying to hand it a file object.

Author: Understood, I rephrased this to explain it needs to be a JSON interface object.

* Always protect your private keys.  Don't leave them lying around unencrypted!

Register with Infura
========================
Before you can connect to a non-local network, you need access to a remote Ethereum node that supports JSON RPC over HTTP.  `Infura <https://infura.io>`__ is one such option.  Once you register and create a project, Infura will provide you URLs with the API key that can be leveraged to access the given network.

Non-local Network Configuration
================================

The connection settings for non-local networks must be defined in ``brownie-config.json``.

First, for each network you want to configure, create a new section in the network.networks section as below:

.. code-block:: javascript
    "networks": {
        .
        .
        "ropsten": {
            "host": "http://ropsten.infura.io/v3/[yourInfuraApiKey]"
        }
        "rinkeby":...
        .
        .
    }

If you want to change the default network that brownie connects to, you need to update the network.default field as below:

.. code-block:: javascript
    "network": {
        "default": "ropsten",
        .
        .
    }

Launching and Connecting
========================

> By default, Brownie will connect to the local network run by `ganache-cli <https://github.com/trufflesuite/ganache-cli>`__.  To connect to a non-local network, you must specify the `--network` flag when launching Brownie as below:

Reviewer: by default, brownie connects to whichever network is given under network.default - using the `--network` flag overrides the config setting

Author: Thanks! I'll clarify that in my next commit. Still working through the content.

::
    $ brownie --network ropsten

Brownie will launch or attach to the client when using any network that includes a ``test-rpc`` dictionary in it's settings.

Each time Brownie is loaded, it will first attempt to connect to the ``host`` address to determine if the RPC client is already active.

Client is Active
----------------

If able to connect to the ``host`` address, Brownie:

* Checks the current block height and raises an Exception if it is greater than zero
* Locates the process listening at the address and attaches it to the ``Rpc`` object
* Takes a snapshot

When Brownie is terminated:

* The RPC client is reverted based on the initial snapshot.

Client is not Active
--------------------

If unable to connect to the ``host`` address, Brownie:

* Launches the client using the ``test-rpc`` command given in the configuration file
* Waits to see that the process loads successfully
* Confirms that it can connect to the new process
* Attaches the process to the ``Rpc`` object

When Brownie is terminated:

* The RPC client and any child processes are also terminated.

> Common Interactions

Reviewer: The `Rpc` object is for interactions with a local client (`ganache-cli`).  If connecting to a non-local network, `Rpc` won't be instantiated so none of this info applies.

Author: Got it. A lot of this content is leftover from when I cloned the test-rpc file to get started. This is my first time doing a project that uses restructured text so needed a quick reference point.
